Mathlib/Data/Real/Irrational.lean

```diff
@@ -93,49 +93,78 @@ theorem irrational_sqrt_of_multiplicity_odd (m : ℤ) (hm : 0 < m) (p : ℕ) [hp
+  · simp_rw [irrational_sqrt_ratCast_iff_of_nonneg hq, and_iff_left hq]
+  · rw [sqrt_eq_zero_of_nonpos (Rat.cast_nonpos.2 hq.le)]
+    simp_rw [not_irrational_zero, false_iff, not_and, not_le, hq, implies_true]
+
+theorem irrational_sqrt_intCast_iff_of_nonneg {z : ℤ} (hz : 0 ≤ z) :
+    Irrational (√z) ↔ ¬IsSquare z := by
+  rw [← Rat.isSquare_intCast_iff, ← irrational_sqrt_ratCast_iff_of_nonneg (mod_cast hz),
+    Rat.cast_intCast]
+
+theorem irrational_sqrt_intCast_iff {z : ℤ} :
+    Irrational (√z) ↔ ¬IsSquare z ∧ 0 ≤ z := by
+  rw [← Rat.cast_intCast, irrational_sqrt_ratCast_iff, Rat.isSquare_intCast_iff, Int.cast_nonneg]
+
+theorem irrational_sqrt_natCast_iff {n : ℕ} : Irrational (√n) ↔ ¬IsSquare n := by
+  rw [← Rat.isSquare_natCast_iff, ← irrational_sqrt_ratCast_iff_of_nonneg n.cast_nonneg,
+    Rat.cast_natCast]
+
+-- See note [no_index around OfNat.ofNat]
+theorem irrational_sqrt_ofNat_iff {n : ℕ} [n.AtLeastTwo] :
+    Irrational (√(no_index (OfNat.ofNat n))) ↔ ¬IsSquare (OfNat.ofNat n) :=
+  irrational_sqrt_natCast_iff
+
 theorem Nat.Prime.irrational_sqrt {p : ℕ} (hp : Nat.Prime p) : Irrational (√p) :=
-  @irrational_sqrt_of_multiplicity_odd p (Int.natCast_pos.2 hp.pos) p ⟨hp⟩ <| by
-    simp [multiplicity.multiplicity_self
-      (mt isUnit_iff_dvd_one.1 (mt Int.natCast_dvd_natCast.1 hp.not_dvd_one))]
+  irrational_sqrt_natCast_iff.mpr hp.not_square
 #align nat.prime.irrational_sqrt Nat.Prime.irrational_sqrt
 
 /-- **Irrationality of the Square Root of 2** -/
 theorem irrational_sqrt_two : Irrational (√2) := by
   simpa using Nat.prime_two.irrational_sqrt
 #align irrational_sqrt_two irrational_sqrt_two
 
+@[deprecated irrational_sqrt_ratCast_iff (since := "2024-06-16")]
 theorem irrational_sqrt_rat_iff (q : ℚ) :
-    Irrational (√q) ↔ Rat.sqrt q * Rat.sqrt q ≠ q ∧ 0 ≤ q :=
-  if H1 : Rat.sqrt q * Rat.sqrt q = q then
-    iff_of_false
-      (not_not_intro
-        ⟨Rat.sqrt q, by
-          rw [← H1, cast_mul, sqrt_mul_self (cast_nonneg.2 <| Rat.sqrt_nonneg q), sqrt_eq,
-            abs_of_nonneg (Rat.sqrt_nonneg q)]⟩)
-      fun h => h.1 H1
-  else
-    if H2 : 0 ≤ q then
-      iff_of_true
-        (fun ⟨r, hr⟩ =>
-          H1 <|
-            (exists_mul_self _).1
-              ⟨r, by
-                rwa [eq_comm, sqrt_eq_iff_mul_self_eq (cast_nonneg.2 H2), ← cast_mul,
-                      Rat.cast_inj] at hr
-                rw [← hr]
-                exact Real.sqrt_nonneg _⟩)
-        ⟨H1, H2⟩
-    else
-      iff_of_false
-        (not_not_intro
-          ⟨0, by
-            rw [cast_zero]
-            exact (sqrt_eq_zero_of_nonpos (Rat.cast_nonpos.2 <| le_of_not_le H2)).symm⟩)
-        fun h => H2 h.2
+    Irrational (√q) ↔ Rat.sqrt q * Rat.sqrt q ≠ q ∧ 0 ≤ q := by
+  rw [irrational_sqrt_ratCast_iff, ne_eq, ← Rat.exists_mul_self]
+  simp only [eq_comm, IsSquare]
 #align irrational_sqrt_rat_iff irrational_sqrt_rat_iff
 
+/--
+This can be used as
+```lean
+unseal Nat.sqrt.iter in
+example : Irrational √24 := by decide
+```
```

Comment:
Maybe mention this in the module docstring, too? (In particular the `unseal` part, as this is not at all obvious...)

Reply:
Done.